Gems/Terrain/Assets/Shaders/Terrain/TerrainDetailHelpers.azsli

```diff
@@ -294,20 +326,55 @@ void GetDebugDetailSurface(inout DetailSurface surface, uint material1, uint mat
 }
 
 //Blend a single detail material sample (with two possible material ids) onto a DetailSurface.
-void BlendDetailMaterial(inout DetailSurface surface, uint material1, uint material2, float blend, float2 detailUv, float weight, float3 macroColor)
+void BlendDetailMaterial(inout DetailSurface surface, in MaterialContext material1, in MaterialContext material2, float blend, float weight, float3 macroColor)
 {
     DetailSurface tempSurface;
-    GetDetailSurfaceForMaterial(tempSurface, material1, detailUv, macroColor);
+    GetDetailSurfaceForMaterial(tempSurface, material1, macroColor);
     WeightDetailSurface(tempSurface, weight * (1.0 - blend));
     AddDetailSurface(surface, tempSurface);
-    if (material2 != 0xFF)
+    if (material2.m_materialId != 0xFF)
     {
-        GetDetailSurfaceForMaterial(tempSurface, material2, detailUv, macroColor);
+        GetDetailSurfaceForMaterial(tempSurface, material2, macroColor);
         WeightDetailSurface(tempSurface, weight * blend);
         AddDetailSurface(surface, tempSurface);
     }
 }
 
+void Swap(inout uint a, inout uint b)
+{
+    uint temp = a;
+    a = b;
+    b = temp;
+}
+
+// Sorter for materials by ids
+void SortMaterialsById(in MaterialContext materialContexts[8], inout uint materialOrder[8])
+{
+    const int sortNetwork[19][2] = {{3, 7}, {2, 6}, {1, 5}, {0, 4}, {5, 7}, {4, 6}, {1, 3}, {0, 2}, {3, 5}, {2, 4}, {6, 7}, {4, 5}, {2, 3}, {0, 1}, {3, 6}, {1, 4}, {5, 6}, {3, 4}, {1, 2}};
```

Comment:
This could really use some comments to help future code maintainers. In particular, an explanation of how to make a new one of these would be nice. :)

Reply:
Generating an optimal sort network is hard, and optimal ones are only known for small numbers of compares. You can optimize for depth (for parallel processing) or total number of compares. It's not quite as simple as "run quicksort and record the comparisons" ... that'll get you something decent but not optimal. I found it quite interesting that you can sort 8 items with 19 compares since 8*log2(8) = 24, so the sort network is about 20% faster. See https://en.wikipedia.org/wiki/Sorting_network for an overview. I used https://demonstrations.wolfram.com/SortingNetworks/ to generate the network for 8 compares. I'll add those links in comments.